plugin/storage/badger/factory.go

```diff
@@ -0,0 +1,179 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package badger
+
+import (
+	"expvar"
+	"flag"
+	"io/ioutil"
+	"os"
+	"time"
+
+	"github.com/dgraph-io/badger"
+	"github.com/spf13/viper"
+	"github.com/uber/jaeger-lib/metrics"
+	"go.uber.org/zap"
+
+	badgerStore "github.com/jaegertracing/jaeger/plugin/storage/badger/spanstore"
+	"github.com/jaegertracing/jaeger/storage/dependencystore"
+	"github.com/jaegertracing/jaeger/storage/spanstore"
+)
+
+var (
+	// ValueLogSpaceAvailable returns the amount of space left on the value log mount point in bytes
+	ValueLogSpaceAvailable *expvar.Int
+	// KeyLogSpaceAvailable returns the amount of space left on the key log mount point in bytes
+	KeyLogSpaceAvailable *expvar.Int
+	// LastMaintenanceRun stores the timestamp of the previous maintenanceRun
+	LastMaintenanceRun *expvar.Int
+	// LastValueLogCleaned stores the timestamp of the previous ValueLogGC run
+	LastValueLogCleaned *expvar.Int
+)
+
+// Factory implements storage.Factory for Badger backend.
+type Factory struct {
+	Options *Options
+	store   *badger.DB
+	cache   *badgerStore.CacheStore
+	logger  *zap.Logger
+
+	tmpDir              string
+	maintenanceInterval time.Duration
+	maintenanceTicker   *time.Ticker
+}
+
+const (
+	defaultTickerInterval time.Duration = 5 * time.Minute
+)
+
+// NewFactory creates a new Factory.
+func NewFactory() *Factory {
+	if ValueLogSpaceAvailable == nil {
+		ValueLogSpaceAvailable = expvar.NewInt("badger_value_log_bytes_available")
+	}
+	if KeyLogSpaceAvailable == nil {
+		KeyLogSpaceAvailable = expvar.NewInt("badger_key_log_bytes_available")
+	}
+	if LastMaintenanceRun == nil {
+		LastMaintenanceRun = expvar.NewInt("badger_storage_maintenance_last_run")
+	}
+	if LastValueLogCleaned == nil {
+		LastValueLogCleaned = expvar.NewInt("badger_storage_valueloggc_last_run")
+	}
+	return &Factory{
+		Options:             NewOptions("badger"),
+		maintenanceInterval: defaultTickerInterval,
+	}
+}
+
+// AddFlags implements plugin.Configurable
+func (f *Factory) AddFlags(flagSet *flag.FlagSet) {
+	f.Options.AddFlags(flagSet)
+}
+
+// InitFromViper implements plugin.Configurable
+func (f *Factory) InitFromViper(v *viper.Viper) {
+	f.Options.InitFromViper(v)
+}
+
+// Initialize implements storage.Factory
+func (f *Factory) Initialize(metricsFactory metrics.Factory, logger *zap.Logger) error {
+	f.logger = logger
+
+	opts := badger.DefaultOptions
+
+	if f.Options.primary.Ephemeral {
+		opts.SyncWrites = false
+		// Error from TempDir is ignored to satisfy Codecov
+		dir, _ := ioutil.TempDir("", "badger")
+		f.tmpDir = dir
+		opts.Dir = f.tmpDir
+		opts.ValueDir = f.tmpDir
+
+		f.Options.primary.KeyDirectory = f.tmpDir
+		f.Options.primary.ValueDirectory = f.tmpDir
```

Comment:
should these all be subdirectories?  Reusing the same one makes assumption about naming that individual components will use.

Reply:
The idea behind these being under different directories is to allow a mix of SSD + HDD combination. The Keys are recommended to be stored to a SSD for better performance, but the values could reside in the HDD also. There's a different access pattern / compaction process in the badger.

Reply:
The idea behind these being under different directories is to allow a mix of SSD + HDD combination. The Keys are recommended to be stored to a SSD for better performance, but the values could reside in the HDD also. There's a different access pattern / compaction process in the badger.